### PR TITLE
Add basic asyncio support

### PR DESCRIPTION
**Pull request by dstandish**
This is meant to be sort of a hello world for asyncio support in airflow.  It will be refined and extended in the future. E.g. probably we would add more config flexibility re the libraries, connect args etc.  But it's good to start simple.

Anyway, ultimately, I think Airflow really needs to go in this direction: in the new REST API, in the new AIP-72 internal API server, in triggers, and ultimately, in the scheduler.

### CODE REVIEW

**Comment by tirkarthi** [comment]
Is this related to AIP-70?

https://cwiki.apache.org/confluence/display/AIRFLOW/%5BWIP%5D+AIP-70+Migrating+to+asynchronous+programming

File: dev/breeze/tests/test_packages.py [hunk 1]
```diff
@@ -221,6 +221,7 @@ def test_get_documentation_package_path():
             """
     "apache-airflow-providers-common-sql>=1.17.0",
     "apache-airflow>=2.8.0",
+    "asyncpg>=0.30.0",
```
Comment (tirkarthi) [inline dev/breeze/tests/test_packages.py:225]: psycopg 3 also has async support but might conflict with psycopg 2 already installed.

https://www.psycopg.org/psycopg3/docs/advanced/async.html

https://docs.sqlalchemy.org/en/20/dialects/postgresql.html#module-sqlalchemy.dialects.postgresql.psycopg

**Comment by dstandish** [comment]
> Is this related to AIP-70?
> 
> https://cwiki.apache.org/confluence/display/AIRFLOW/%5BWIP%5D+AIP-70+Migrating+to+asynchronous+programming

I think you know the answer.  The AIP deals with asyncio, so yes strictly speaking related in that sense.  But not "part of".  That AIP has been in draft a long time not sure actively worked on.  Anyway, I don't think what I'm doing here really requires an AIP.

File: dev/breeze/tests/test_packages.py (hunk 1 above)
Comment (dstandish) [reply dev/breeze/tests/test_packages.py:225]: > psycopg 3 also has async support but might conflict with psycopg 2 already installed.
> 
> https://www.psycopg.org/psycopg3/docs/advanced/async.html
> 
> https://docs.sqlalchemy.org/en/20/dialects/postgresql.html#module-sqlalchemy.dialects.postgresql.psycopg

True, yeah.  We could in future look at allowing diff libs.

**Comment by tirkarthi** [comment]
There is a similar draft PR open as POC for it though not active. So just wanted to confirm. Thanks for the details.

https://github.com/apache/airflow/pull/36504

**Review by pierrejeambrun — APPROVED** [review]
Really exciting.

Do we want an `@provide_async_session` as a follow up ? (I think Ash wanted to kill that  decorator at some point, even if I find it kind of useful actually)

File: dev/breeze/tests/test_packages.py (hunk 1 above)
Comment (ashb) [reply dev/breeze/tests/test_packages.py:225]: Benchmark, so take with a pinch of salt, but https://pypi.org/project/asyncpg/

> In our testing asyncpg is, on average, 5x faster than psycopg3.
>
> 

That's def a big enough of a change that we should see if we notice any difference

**Comment by ashb** [comment]
@pierrejeambrun A FastAPI dep to give an async session would be good, but I think anything in the models etc should be passed an explicit session.

File: airflow/settings.py [hunk 2]
```diff
@@ -199,13 +209,25 @@ def load_policy_plugins(pm: pluggy.PluginManager):
     pm.load_setuptools_entrypoints("airflow.policy")
 
 
+def _get_async_conn_uri_from_sync(sync_uri):
+    scheme, rest = sync_uri.split(":", maxsplit=1)
```
Comment (ashb) [inline airflow/settings.py:213]: SQLA has stuff built in to parse URLs/schemas. Maybe we should use that instead?

Comment (ashb) [reply airflow/settings.py:213]: https://docs.sqlalchemy.org/en/14/core/engines.html#sqlalchemy.engine.URL

File: airflow/settings.py [hunk 3]
```diff
@@ -441,6 +463,9 @@ def configure_orm(disable_connection_pool=False, pool_class=None):
 
     global Session
     global engine
+    global async_engine
+    global create_async_session
```
Comment (ashb) [inline airflow/settings.py:467]:
```suggestion
```

**Comment by potiuk** [comment]
NICE! Yeah. I really like to see it working with mysql compatibility and benchmarks :package: 

(because of course MySQL is our beloved database)

File: airflow/settings.py (hunk 3 above)
Comment (dstandish) [reply airflow/settings.py:467]: why @ashb ?

Comment (dstandish) [reply airflow/settings.py:467]: i think it's necessary ... it doesn't get set until configure_orm is called

**Comment by dstandish** [comment]
> Really exciting.
> 
> Do we want an `@provide_async_session` as a follow up ? (I think Ash wanted to kill that decorator at some point, even if I find it kind of useful actually)

i would say let's wait and see how / where / when we need it

**Comment by omkar-foss** [comment]
Lovely! Thank you for adding this, will try to start using this soon :)

File: .github/workflows/basic-tests.yml [hunk 4]
```diff
@@ -196,8 +196,14 @@ jobs:
...
+            breeze release-management prepare-provider-packages \
+              fab \
+              standard \
+              common.sql \
+              sqlite \
+              --package-format wheel \
+              --skip-tag-check \
+              --version-suffix-for-pypi dev0
             pip install . dist/apache_airflow_providers_fab-*.whl \
             dist/apache_airflow_providers_standard-*.whl dist/apache_airflow_providers_common_sql-*.whl
```
Comment (kaxil) [inline .github/workflows/basic-tests.yml]:
```suggestion
            dist/apache_airflow_providers_standard-*.whl dist/apache_airflow_providers_common_sql-*.whl dist/apache_airflow_providers_sqlite-*.whl
```

File: tests/utils/test_session.py [hunk 5]
```diff
@@ -53,3 +55,13 @@ def test_provide_session_with_kwargs(self):
...
         assert wrapper(session=session) is session
+
+    @pytest.mark.asyncio
+    async def test_async_session(self):
+        from airflow.settings import create_async_session
+
+        session = create_async_session()
+        session.add(Log(event="hihi1234"))
+        await session.commit()
+        l = await session.scalar(select(Log).where(Log.event == "hihi1234").limit(1))  # noqa: E741
```
Comment (kaxil) [inline tests/utils/test_session.py:66]: Can avoid this noqa by just change the var name :)

File: tests/utils/test_session.py [hunk 6]
```diff
@@ -53,3 +55,13 @@ def test_provide_session_with_kwargs(self):
...
+        assert l.event == "hihi1234"
```
Comment (kaxil) [inline tests/utils/test_session.py:67]:
```suggestion
        log_result = await session.scalar(select(Log).where(Log.event == "hihi1234").limit(1))  # noqa: E741
        assert log_result.event == "hihi1234"
```

File: tests/utils/test_session.py (hunk 5 above)
Comment (dstandish) [reply tests/utils/test_session.py:66]: we can, but i did it just to spite the nanny state :)

File: tests/utils/test_session.py (hunk 6 above)
Comment (dstandish) [reply tests/utils/test_session.py:67]: i can do this in a followup kaxil if it pleases you. don't want to wait for 99 ci checks to run agaih 😬

File: tests/utils/test_session.py (hunk 5 above)
Comment (kaxil) [reply tests/utils/test_session.py:66]: :D

File: tests/utils/test_session.py (hunk 6 above)
Comment (kaxil) [reply tests/utils/test_session.py:67]: haha I definitely understand -- and its a nit so will leave it up to you :)

**Pull request merged**